README.rst
Wagtail JsonSchema Forms
========================

.. image:: https://api.travis-ci.org/APSL/wagtail-jsonschema-forms.svg?branch=master
    :target: https://travis-ci.org/APSL/wagtail-jsonschema-forms

Wagtail JsonSchema Forms is a simple Django app to Generate JsonSchema Forms from Wagtail's forms and validate it.

This app allows the generation of a JsonSchema form from a Wagtail's form in order to use JsonSchema technology.
In addition, it allows you to parse the response and validate it. After the validation, the wagtail form is
submitted.

For example, you can retrieve the jsonschema from through an api, then validate the submitted data and finally
submit de wagtail form.

Installation
------------

You can install "wagtail_jsonschema_forms" via `pip`_ from `APSL GitHub`_:

* pip install git+https://github.com/APSL/wagtail_jsonschema_forms.git#egg=wagtail_jsonschema_forms

And then you have to add :code:`wagtail_jsonschema_forms` to your :code:`INSTALLED_APPS`.


Utilities provided
------------------
This app offers two python classes to generate the jsonschema and to submit the json response.

JsonSchemaGenerator
~~~~~~~~~~~~~~~~~~~

This first class has a method :code:`wagtail_form_to_json_schema` that goes over
all the fields of the form and generate its jsonschema representation, if they are suported.

Every suported field type has a method that generate the representation, which is called during the generation.
If the field is required, it will be indicated too.


JsonSchemaParser
~~~~~~~~~~~~~~~~

This second class provides two main methods:

* :code:`validate_json`:
    method that validates the json recieved under the schema a return a dict with the possible errors.
    If there are no error, a empty dict will be returned.
* :code:`process_form_submission`:
    method that accepts form instance with submitted data and page and creates submission instance.
    Inside of this method, another one is called to format the data submited in order to create the
    submission instance.


Usage
-----
As was explained before, this django app generates jsonschema forms from Wagtail's form and validate it.
Wagtail's forms offers a very rich set of form field types. However, not all of them are suported on
JsonSchema and has not been covered on this app. For that reason, when you declare your :code:`FormField`
class on your models.py you should override the field_type field and the :code:`FORM_FIELD_CHOICES` variable:

.. code-block:: python

    FORM_FIELD_CHOICES = (
        ('singleline', _('Single line text')),
        ('multiline', _('Multi-line text')),
        ('email', _('Email')),
        ('number', _('Number')),
        ('url', _('URL')),
        ('checkbox', _('Checkbox')),
        ('dropdown', _('Drop down')),
        ('date', _('Date')),
        ('datetime', _('Date/time')),
      )

    field_type = models.CharField(verbose_name=_('field type'), max_length=16, choices=FORM_FIELD_CHOICES)

There are two ways to use wagtail_json_schema_forms: you can choose between using the mixins or the abstract
models provided.

Mixins
~~~~~~

If you choose to use the mixins, the first thing thant you have to do is to add a new :code:`TextField` atribute
to your wagtail form page called 'json_schema'. This field will be used to save the JsonSchema representation
of the form.

Two mixins are provided: :code:`JsonSchemaFormMixin` and :code:`JsonSchemaEmailFormMixin`, which inherit from the first one.

:code:`JsonSchemaFormMixin` overrides the save method in order to generate the json schema before every time the
wagtail page is saved. This mixin uses the generator class explained above to prepare the jsonschema from the form.

It's also overrided the method :code:`process_form_submission`, where it's checked if the form has been submited as
a usual wagtail form or using jsonschema. If jsonschema has been used for submition, the parser class is used
to register the form submission.

As said, the :code:`JsonSchemaEmailFormMixin` inherits from :code:`JsonSchemaFormMixin`. The diference between them is in
the process of the form submission. This second mixin tries to send an email after the form submission, if the
:code:`to_address` attribute of the wagtail page has been filed.


Abstract Models
~~~~~~~~~~~~~~~

Again, there are two abstract models provided to allow to add the jsonschema functionality to a plain WagtailForm
and a EmailWagtailForm: :code:`AbstractJsonSchemaForm` and :code:`AbstractJsonSchemaEmailForm`.

These abstract models inherit from the mixins explained above and the :code:`AbstractForm` or :code:`AbstractEmailForm`.
They also add the new :code:`TextField` atribute required. Then, Wagtail Pages implementing a form that need to
generate a JsonSchema (and to send and email) should inherit from it.



Contributing
------------

Contributions are very welcome. Please open a pull request or `file an issue`_.
Tests will be ready as soon as posible, please ensure the coverage at least stays the same
before you submit a pull request.

License
-------

Distributed under the terms of the `MIT`_ license, "wagtail_jsonschema_forms" is free and open source software


Issues
------

If you encounter any problems, please `file an issue`_ along with a detailed description.

.. _`pip`: https://pypi.python.org/pypi/pip/
.. _`APSL GitHub`: https://github.com/APSL/wagtail_jsonschema_forms
.. _`file an issue`: https://github.com/APSL/wagtail_jsonschema_forms/issues
.. _`MIT`: http://opensource.org/licenses/MIT

CHANGELOG
Changelog
=========

0.1.0 (2017-10-13)
------------------
* Initial release


tests/conftest.py
import json
from collections import OrderedDict
import pytest
from model_mommy import mommy
from wagtail.tests.testapp.models import FormPage, FormField
from wagtail.wagtailcore.models import Page


@pytest.mark.django_db()
@pytest.fixture()
def form_page(admin_user):
    root_page = Page.objects.first()
    form = FormPage()
    form.title = "form"
    form.slug = "form"
    form.live = True
    form.owner = admin_user
    root_page.add_child(instance=form)
    form.save()
    return form


@pytest.mark.django_db()
@pytest.fixture()
def singleline(form_page):
    return mommy.make(FormField, page=form_page, field_type="singleline", label="singleline",
                      help_text="singleline", default_value="singleline", required=True)


@pytest.fixture()
def jsonschema_singleline():
    return {
        "type": "string",
        "description": "singleline",
        "maxLength": 150,
        "default": "singleline"
    }


@pytest.mark.django_db()
@pytest.fixture()
def multiline(form_page):
    return mommy.make(FormField, page=form_page, field_type="multiline", label="multiline", default_value="multiline",
                      required=False)


@pytest.fixture()
def jsonschema_multiline():
    return {
        "type": "string",
        "description": "",
        "maxLength": 500,
        "default": "multiline"
    }


@pytest.mark.django_db()
@pytest.fixture()
def date(form_page):
    return mommy.make(FormField, page=form_page, field_type="date", label="date", help_text="date", required=True)


@pytest.fixture()
def jsonschema_date():
    return {
        "type": "string",
        "description": "date",
        "format": "date",
        "default": ""
    }


@pytest.mark.django_db()
@pytest.fixture()
def datetime(form_page):
    return mommy.make(FormField, page=form_page, field_type="datetime", label="datetime", help_text="datetime",
                      required=False)


@pytest.fixture()
def jsonschema_datetime():
    return {
        "type": "string",
        "description": "datetime",
        "format": "date-time",
        "default": ""
    }


@pytest.mark.django_db()
@pytest.fixture()
def email(form_page):
    return mommy.make(FormField, page=form_page, field_type="email", label="email", default_value="email",
                      required=False)


@pytest.fixture()
def jsonschema_email():
    return {
        "type": "string",
        "description": "",
        "maxLength": 100,
        "format": "email",
        "default": "email"
    }


@pytest.mark.django_db()
@pytest.fixture()
def url(form_page):
    return mommy.make(FormField, page=form_page, field_type="url", label="url", required=False)


@pytest.fixture()
def jsonschema_url():
    return {
        "type": "string",
        "description": "",
        "maxLength": 100,
        "format": "uri",
        "default": ""
    }


@pytest.mark.django_db()
@pytest.fixture()
def number(form_page):
    return mommy.make(FormField, page=form_page, field_type="number", label="number", required=True)


@pytest.fixture()
def jsonschema_number():
    return {
        "type": "number",
        "description": "",
        "default": ""
    }


@pytest.mark.django_db()
@pytest.fixture()
def dropdown(form_page):
    return mommy.make(FormField, page=form_page, field_type="dropdown", label="dropdown", help_text="dropdown",
                      choices="dropdown,dropdown2,dropdown3", required=False)


@pytest.fixture()
def jsonschema_dropdown():
    return {
        "type": "string",
        "description": "dropdown",
        "default": "",
        "enum": ["dropdown", "dropdown2", "dropdown3"]
    }


@pytest.mark.django_db()
@pytest.fixture()
def checkbox(form_page):
    return mommy.make(FormField, page=form_page, field_type="checkbox", label="checkbox", required=True)


@pytest.fixture()
def jsonschema_checkbox():
    return {
        "type": "boolean",
        "description": "",
        "default": "",
        "format": "checkbox"
    }


@pytest.mark.django_db()
@pytest.fixture()
def wagtail_form(checkbox, dropdown, number, url, email, datetime, date, multiline, singleline):
    return singleline.page


@pytest.fixture()
def jsonschema_generated(jsonschema_singleline, jsonschema_multiline, jsonschema_date, jsonschema_datetime,
                         jsonschema_email, jsonschema_url, jsonschema_number, jsonschema_dropdown, jsonschema_checkbox):

    properties = OrderedDict()
    properties["checkbox"] = jsonschema_checkbox
    properties["dropdown"] = jsonschema_dropdown
    properties["number"] = jsonschema_number
    properties["url"] = jsonschema_url
    properties["email"] = jsonschema_email
    properties["datetime"] = jsonschema_datetime
    properties["date"] = jsonschema_date
    properties["multiline"] = jsonschema_multiline
    properties["singleline"] = jsonschema_singleline

    json_schema = OrderedDict()
    json_schema["title"] = "form"
    json_schema["type"] = "object"
    json_schema["properties"] = properties
    json_schema["required"] = ["checkbox", "number", "date", "singleline"]
    return json.dumps(json_schema)


########################################################################################################################
# Parser fixtures
########################################################################################################################

@pytest.mark.django_db()
@pytest.fixture()
def single_line(form_page):
    return mommy.make(FormField, page=form_page, field_type="singleline", label="single line",
                      help_text="singleline", default_value="singleline", required=True)


@pytest.mark.django_db()
@pytest.fixture()
def simple_wagtail_form(single_line):
    return single_line.page


@pytest.fixture()
def simple_form_json_correct():
    return {"single line": "Test value"}


@pytest.fixture()
def simple_form_json_incorrect():
    return {"single line": 31351351}


@pytest.fixture()
def json_submited():
    return {"single line": "Test value"}


@pytest.fixture()
def json_ready_to_save():
    return '{"single-line": "Test value"}'


tests/testapp/settings.py
import os

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SECRET_KEY = 'changemepliz'
DEBUG = True
ALLOWED_HOSTS = "*"

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'wagtail.wagtailsites',
    'wagtail.wagtailforms',
    'wagtail.wagtailusers',
    'wagtail.wagtaildocs',
    'wagtail.wagtailimages',
    'wagtail.wagtailcore',
    'wagtail.contrib.modeladmin',
    'wagtail.tests.testapp',
    'taggit',
    'wagtail_jsonschema_forms'
)

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'tests.testapp.wsgi.application'

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'tests/testapp/wagtail_jsonschema_forms.db',
        'USER': '',
        'PASSWORD': '',
    }
}

WAGTAIL_SITE_NAME = 'Wagtail JsonSchema Test Site'


tests/unit/test_jsonschema_parser.py

import json

from wagtail_jsonschema_forms.jsonschema_parser import JsonSchemaParser
from wagtail_jsonschema_forms.jsonschema_generator import JsonSchemaGenerator


class TestJsonschemaParser:

    def test_validate_json_correct(self, simple_form_json_correct, simple_wagtail_form):
        form_schema = JsonSchemaGenerator.wagtail_form_to_json_schema(simple_wagtail_form)
        form_schema = json.loads(form_schema)
        errors = JsonSchemaParser.validate_json(simple_form_json_correct, form_schema)
        assert errors == {}

    def test_validate_json_incorrect(self, simple_form_json_incorrect, simple_wagtail_form):
        form_schema = JsonSchemaGenerator.wagtail_form_to_json_schema(simple_wagtail_form)
        form_schema = json.loads(form_schema)
        errors = JsonSchemaParser.validate_json(simple_form_json_incorrect, form_schema)
        assert "single line" in errors

    def test_prepare_form_data(self, json_submited, json_ready_to_save):
        assert JsonSchemaParser.prepare_form_data(json_submited) == json_ready_to_save

    def test_process_form_submission(self, simple_wagtail_form, json_submited):
        JsonSchemaParser.process_form_submission(simple_wagtail_form, json_submited)
        assert simple_wagtail_form.get_submission_class().objects.all()


wagtail_jsonschema_forms/mixins.py

from wagtail.wagtailadmin.utils import send_mail
from wagtail_jsonschema_forms.jsonschema_generator import JsonSchemaGenerator
from wagtail_jsonschema_forms.jsonschema_parser import JsonSchemaParser


class JsonSchemaFormMixin:

    def save(self, *args, **kwargs):
        self.json_schema = self.generate_json_schema()
        super().save(*args, **kwargs)

    def generate_json_schema(self):
        """
        Function that generates the json schema representation of the form
        """
        return JsonSchemaGenerator.wagtail_form_to_json_schema(self)

    def process_form_submission(self, form):
        """
        Function that check the way that the form has been submited
        :param form: WagtailForm or dict
        :return:
        """
        if hasattr(form, "data"):
            super().process_form_submission(form)
        else:
            self.process_json_form_submission(form)

    def process_json_form_submission(self, json_submited):
        """
        Function that process json form submission
        :param json_submited: dict
        :return:
        """
        JsonSchemaParser.process_form_submission(self, json_submited)


class JsonSchemaEmailFormMixin(JsonSchemaFormMixin):

    def process_json_form_submission(self, json_submited):
        """
        Function that process json form submission and send an email from a json representation of the submitted data
        :param json_submited: dict
        :return:
        """
        JsonSchemaParser.process_form_submission(self, json_submited)
        self.send_mail_from_json(json_submited)

    def send_mail_from_json(self, form_fields):
        """
        Function that send a email from a json representation of the post data of the form.
        Based on wagtail.wagtailforms.models.AbstractEmailForm.send_mail
        :param form_fields: dict
        :return:
        """
        if self.to_address:
            addresses = [x.strip() for x in self.to_address.split(',')]
            content = []
            for field_key, value in form_fields.items():
                if isinstance(value, list):
                    value = ', '.join(value)
                content.append('{}: {}'.format(field_key.capitalize(), value))
            content = '\n'.join(content)
            send_mail(self.subject, content, addresses, self.from_address)


wagtail_jsonschema_forms/models.py
from django.db import models
from wagtail.wagtailforms.models import AbstractForm, AbstractEmailForm

from wagtail_jsonschema_forms.mixins import JsonSchemaFormMixin, JsonSchemaEmailFormMixin


class AbstractJsonSchemaForm(JsonSchemaFormMixin, AbstractForm):
    """
    A Form Page that generates a JSON Schema representation of the fields and save it to the model.
    Pages implementing a form that need to generate JSON Schema should inherit from it
    """

    json_schema = models.TextField(help_text=u"Form json Schema representation", blank=True)

    class Meta:
        abstract = True


class AbstractJsonSchemaEmailForm(JsonSchemaEmailFormMixin, AbstractEmailForm):
    """
    A Form Page that generates a JSON Schema representation of the fields and save it to the model. It also send email.
    Pages implementing a form that need to generate JSON Schema and need to send an email should inherit from it
    """

    json_schema = models.TextField(help_text=u"Form json Schema representation", blank=True)

    class Meta:
        abstract = True


wagtail_jsonschema_forms/jsonschema_parser.py

from django.core.serializers.json import DjangoJSONEncoder
from django.utils.six import text_type
from django.utils.text import slugify

import json
from jsonschema.validators import validator_for
from unidecode import unidecode


class JsonSchemaParser:

    @staticmethod
    def validate_json(json_form, schema, *args, **kwargs):
        """
        Function that validates the json recieved under the schema and returns a dict with the errors, if there are any
        :param json_form: dict
        :param schema: dict
        :return dict
        """
        cls = validator_for(schema)
        cls.check_schema(schema)
        json_errors = cls(schema, *args, **kwargs).iter_errors(json_form)
        errors = {}
        for error in json_errors:
            # errors can be from fields validation or required camps
            form_field = error.path[0] if error.path else error.validator_value[0]
            errors[form_field] = error.message
        return errors

    @staticmethod
    def process_form_submission(form, json_submited):
        """
        Function that accepts form instance with submitted data and page and creates submission instance.
        :param form: Form Page
        :param json_submited: str
        """
        form.get_submission_class().objects.create(form_data=JsonSchemaParser.prepare_form_data(json_submited),
                                                   page=form)

    @staticmethod
    def prepare_form_data(json_submited):
        """
        Function that prepare json field keys to the submition
        :param json_submited: dict
        :return str
        """
        form_data = {}
        for field_key, value in json_submited.items():
            # same conversion as clean_name in wagtail.wagtailforms.models.AbstractFormField
            form_data[str(slugify(text_type(unidecode(field_key))))] = value
        return json.dumps(form_data, cls=DjangoJSONEncoder)


wagtail_jsonschema_forms/jsonschema_generator.py
from collections import OrderedDict
import json


class JsonSchemaGenerator:

    FIELD_TYPES = {
        "singleline": "generate_singleline",
        "multiline": "generate_multiline",
        "date": "generate_date",
        "datetime": "generate_datetime",
        "email": "generate_email",
        "url": "generate_url",
        "number": "generate_number",
        "dropdown": "generate_dropdown",
        "checkbox": "generate_checkbox",
    }

    @staticmethod
    def generate_singleline(field):
        return {
            "type": "string",
            "description": field.help_text,
            "maxLength": 150,
            "default": field.default_value
        }

    @staticmethod
    def generate_multiline(field):
        return {
            "type": "string",
            "description": field.help_text,
            "maxLength": 500,
            "default": field.default_value
        }

    @staticmethod
    def generate_email(field):
        return {
            "type": "string",
            "description": field.help_text,
            "maxLength": 100,
            "format": "email",
            "default": field.default_value
        }

    @staticmethod
    def generate_date(field):
        return {
            "type": "string",
            "description": field.help_text,
            "format": "date",
            "default": field.default_value
        }

    @staticmethod
    def generate_datetime(field):
        return {
            "type": "string",
            "description": field.help_text,
            "format": "date-time",
            "default": field.default_value
        }

    @staticmethod
    def generate_url(field):
        return {
            "type": "string",
            "description": field.help_text,
            "maxLength": 100,
            "format": "uri",
            "default": field.default_value
        }

    @staticmethod
    def generate_number(field):
        return {
            "type": "number",
            "description": field.help_text,
            "default": field.default_value
        }

    @staticmethod
    def generate_dropdown(field):
        return {
            "type": "string",
            "description": field.help_text,
            "default": field.default_value,
            "enum": field.choices.split(",")
        }

    @staticmethod
    def generate_checkbox(field):
        return {
            "type": "boolean",
            "description": field.help_text,
            "default": field.default_value,
            "format": "checkbox"
        }

    @staticmethod
    def wagtail_form_to_json_schema(form):
        """
        This function expects a wagtail form and return a Json Schema representation
        :param form: wagtail form
        :return: dict
        """
        json_schema = OrderedDict()
        json_schema["title"] = form.slug
        json_schema["type"] = "object"

        properties = OrderedDict()
        required = []
        for field in form.form_fields.all():
            field_type = field.field_type
            if field_type in JsonSchemaGenerator.FIELD_TYPES:
                properties[field.label] = getattr(JsonSchemaGenerator,
                                                  JsonSchemaGenerator.FIELD_TYPES[field_type])(field)

                if field.required:
                    required.append(field.label)

        json_schema["properties"] = properties
        json_schema["required"] = required
        return json.dumps(json_schema)
